tests/unit/test_serialization.py

```diff
@@ -34,6 +36,70 @@ def from_msgpack(data):
     return msgpack.unpackb(data)
 
 
+def is_non_compact_list(data_entry):
+    return isinstance(data_entry, dict)
+
+
+def is_serialized_data_type_deducible(serialized_input, data_filter) -> bool:
+    """Find out if deserialization of serialized_input would contain atleast one component with row based data"""
+    serialized_input_data = serialized_input["data"]
+    if not serialized_input_data or data_filter is Ellipsis:
+        return False
+    if data_filter is None:
+        return True
+
+    if not serialized_input["is_batch"]:
+        components = set(serialized_input_data.keys())
+    else:
+        components = set()
+        for scenario in serialized_input_data:
+            components.update(scenario.keys())
+
+    return any(data_filter[comp] is None for comp in components.intersection(data_filter.keys()))
+
+
+def is_columnar_filter(data_filter, component):
+    """A function to find if a data_filter will give out row or columnar format for a component"""
+    if data_filter is None:
+        return False
+    if data_filter is Ellipsis:
+        return True
+    return data_filter[component] is not None
+
+
+def is_filtered_out(data_filter, component, attribute=None):
+    if data_filter is None or data_filter is Ellipsis:
+        return False
+    if attribute is None:
+        return component not in data_filter
+    return not (
+        data_filter[component] is None or data_filter[component] is Ellipsis or attribute in data_filter[component]
+    )
+
+
+def is_sparse_data_input(serialized_input_data, component):
+    """Checks if serialized_input_data will be of sparse format after deserialization.
+    If there are uneven ids for scenarios in the serialized input, that would result in sparse data"""
```

Comment:
I would argue that any uneven attribute, not just ids, leads to sparse data.

Reply:
Yes indeed. The goal is to deduce if the input to tests, which is serialized in this case, would be converted to a sparse or a dense dataset.
The logic is only applicable for this test file.

And when its uneven content inside each batch, its sparse dataset. "id" is the common attribute in all cases here, hence used "id" to differentiate between them.